lib/host/gen_helpers.js

```diff
@@ -146,15 +172,43 @@ const genHelpers = (state = {}, adapter) => {
       await state.ev.emit('notify', gid, level, message)
     },
 
-    subscribeDataChannels: (state = {}) => {
+    /**
+     *
+     * @param {Object} state
+     * @param {Object} options
+     * @param {boolean} [options.fireAndForget]
+     * @param {number} [options.timeout]
+     * @returns Promise
+     */
+    subscribeDataChannels: (state = {}, options = {}) => {
       const { channels = [], gid, connection } = state
 
-      channels.forEach((ch) => {
-        debug('subscribing to channel %j [AO gid %d]', ch, gid)
+      return Promise.all(
+        channels.map(({ channel, filter: payload }) => {
+          debug('subscribing to channel %j (filter: %) [AO gid %d]', channel, payload, gid)
 
-        // TODO: optional await
-        adapter.subscribe(connection, ch.channel, ch.filter)
-      })
+          adapter.subscribe(connection, channel, payload)
+
+          if (options.fireAndForget) {
+            return Promise.resolve()
+          }
+
+          const { promise, resolve, reject } = flatPromise()
+          const timeout = options.timeout &&
+            setTimeout(() => reject(new Error(`Subscription to channel '${channel}' timed out`)), options.timeout)
```

Comment:
may be delete `pendingSubscriptionResponses[eid]` on timeout function as well? Otherwise, `pendingSubscriptionResponses` will have the data there always if not subscribed.

Reply:
done!